### ISSUE TITLE
build_type not found while building test_package/

### ISSUE THREAD

**Issue by matlo607** [conan-io/conan#2500]
I created a recipe in which I had to change `build_type` to `RelWithDebInfo` on windows.
```python
   def build(self):
        build_type = self.settings.build_type
        if self.settings.compiler == "Visual Studio" and self.settings.build_type == "Release":
            build_type = "RelWithDebInfo"

        cmake = CMake(self, build_type=build_type)
        cmake.definitions['CMAKE_INSTALL_PREFIX'] = '{}'.format(self.package_folder)
        ...
```
I did the same inside the test_package.
```python
import os
from conans import ConanFile, CMake, RunEnvironment, tools

username = os.getenv("CONAN_USERNAME", "matthieu")
channel = os.getenv("CONAN_CHANNEL", "testing")

class ToolsTestConan(ConanFile):
    settings = "os", "compiler", "build_type", "arch"
    requires = ( 
        "tools/0.3.2@%s/%s" % (username, channel),
        "googletest/1.8.0@%s/%s" % (username, channel)
    )   
    generators = "cmake"
    _cmake = None

    def build(self):
        build_type = self.settings.build_type
        if self.settings.compiler == "Visual Studio" and self.settings.build_type == "Release":
            build_type = "RelWithDebInfo"

        self._cmake = CMake(self, build_type=build_type)
        self._cmake.configure(args=['-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON'])
        self._cmake.build()

    def imports(self):
        pass

    def test(self):
        env_tests = RunEnvironment(self)
        with tools.environment_append(env_tests.vars):
            self._cmake.test()
```

When I ran the build on Windows, the package was built perfectly.
However, when I try to build the recipe on Linux with `conan create -p gcc-49 . matthieu/testing --build=missing`, the library is built but when conan arrives at the test_package step, I get the following error :
```
tools/0.3.2@matthieu/testing: Package '185310967a7b09ff8cd7b5be4ee6995d474bbce7' built
tools/0.3.2@matthieu/testing: Build folder /home/matt/.conan/data/tools/0.3.2/matthieu/testing/build/185310967a7b09ff8cd7b5be4ee6995d474bbce7
tools/0.3.2@matthieu/testing: Generated conaninfo.txt
tools/0.3.2@matthieu/testing: Generated conanbuildinfo.txt
tools/0.3.2@matthieu/testing: Generating the package
tools/0.3.2@matthieu/testing: Package folder /home/matt/.conan/data/tools/0.3.2/matthieu/testing/package/185310967a7b09ff8cd7b5be4ee6995d474bbce7
tools/0.3.2@matthieu/testing: Calling package()
tools/0.3.2@matthieu/testing package(): WARN: No files copied!
tools/0.3.2@matthieu/testing: Package '185310967a7b09ff8cd7b5be4ee6995d474bbce7' created
tools/0.3.2@matthieu/testing (test package): Generator cmake created conanbuildinfo.cmake
tools/0.3.2@matthieu/testing (test package): Generator txt created conanbuildinfo.txt
tools/0.3.2@matthieu/testing (test package): Generated conaninfo.txt
tools/0.3.2@matthieu/testing (test package): Running build()
ERROR: PROJECT: Error in build() method, line 17
	build_type = self.settings.build_type
	ConanException: 'settings.build_type' doesn't exist
'settings' possible configurations are ['arch', 'compiler', 'os']
```

Running `conan create -p gcc-49 . matthieu/testing --build=missing -s build_type=Release` makes the build succeed.
I seems that the field `build_type` is not set in test_package. I tried with other fields like `compiler` or `os` and they are correctly set.

**Comment by memsharded** [comment]
Could you please try this?:

```python
if self.settings.compiler == "Visual Studio" and self.settings.build_type == "Release":
    build_type = "RelWithDebInfo"
else:
    build_type = self.settings.build_type
```

**Comment by matlo607** [comment]
It still fails.

**Comment by memsharded** [comment]
I have tried it, but I couldn't make it fail in the same way (also in Linux).
If you could please try to reduce the code (like removing unnecessary requires) to the minimum, and put the full example in a github repo, with the steps you are following, that would really help.

Some quick observations:

- ``cmake.definitions['CMAKE_INSTALL_PREFIX'] = '{}'.format(self.package_folder)`` might not be necessary. This is already automatically managed by conan
- Instead of ``self._cmake.configure(args=['-DCMAKE_VERBOSE_MAKEFILE:BOOL=ON'])``, you can try: ``cmake.verbose=True``
- I have submitted a new issue: https://github.com/conan-io/conan/issues/2501, that will simplify the test_package code

**Comment by matlo607** [comment]
I managed to reproduce the issue.

`git clone https://github.com/lasote/conan-gtest.git`
`git am 0001-Bug-with-build_type.patch`
```
From 8657b3344132d3217bb2c0025348a97a91d331da Mon Sep 17 00:00:00 2001
From: Matthieu <foo@foo.com>
Date: Tue, 20 Feb 2018 17:51:22 +0100
Subject: [PATCH] Bug with build_type

---
 conanfile.py              | 6 +++++-
 test_package/conanfile.py | 6 +++++-
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/conanfile.py b/conanfile.py
index f35d065..197508e 100644
--- a/conanfile.py
+++ b/conanfile.py
@@ -34,9 +34,13 @@ class GTestConan(ConanFile):
         os.unlink(zip_name)
 
     def build(self):
+        build_type = self.settings.build_type
+        if self.settings.compiler == "Visual Studio" and self.settings.build_type == "Release":
+            build_type = "RelWithDebInfo"
+
         files.mkdir("_build")
         with tools.chdir("_build"):
-            cmake = CMake(self)
+            cmake = CMake(self, build_type=build_type)
             if self.settings.compiler == "Visual Studio" and "MD" in str(self.settings.compiler.runtime):
                 cmake.definitions["gtest_force_shared_crt"] = "ON"
             if self.options.shared:
diff --git a/test_package/conanfile.py b/test_package/conanfile.py
index aebbffe..3e85799 100644
--- a/test_package/conanfile.py
+++ b/test_package/conanfile.py
@@ -11,7 +11,11 @@ class DefaultNameConan(ConanFile):
     requires = "gtest/1.8.0@lasote/stable"
         
     def build(self):
-        cmake = CMake(self)
+        build_type = self.settings.build_type
+        if self.settings.compiler == "Visual Studio" and self.settings.build_type == "Release":
+            build_type = "RelWithDebInfo"
+
+        cmake = CMake(self, build_type=build_type)
         cmake.configure()
         cmake.build()
 
-- 
2.7.4
```

Output:
```
conan create -p gcc-49 . lasote/testing
gtest/1.8.0@lasote/testing: Exporting package recipe
gtest/1.8.0@lasote/testing export: Copied 1 '.txt' files: CMakeLists.txt
gtest/1.8.0@lasote/testing: The stored package has not changed
gtest/1.8.0@lasote/testing (test package): Installing /home/matt/Documents/conan-mx/recipes/conan-gtest-lasote/test_package/conanfile.py
Requirements
    gtest/1.8.0@lasote/testing from local
Packages
    gtest/1.8.0@lasote/testing:0f715b9d941e4c9bebe4c1829b733a968bf42836

gtest/1.8.0@lasote/testing: WARN: Forced build from source
gtest/1.8.0@lasote/testing: Building your package in /home/matt/.conan/data/gtest/1.8.0/lasote/testing/build/0f715b9d941e4c9bebe4c1829b733a968bf42836
gtest/1.8.0@lasote/testing: Copying sources to build folder
gtest/1.8.0@lasote/testing: Generator cmake created conanbuildinfo.cmake
gtest/1.8.0@lasote/testing: Calling build()
-- The C compiler identification is GNU 4.9.4
-- The CXX compiler identification is GNU 4.9.4
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
Conan Gtest Wrapped CMake
-- Conan: called by CMake conan helper
-- Conan: Checking correct version: 4.9
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Conan: C++ stdlib: libstdc++11
-- Found PythonInterp: /usr/bin/python (found version "2.7.12") 
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Configuring done
-- Generating done
-- Build files have been written to: /home/matt/.conan/data/gtest/1.8.0/lasote/testing/build/0f715b9d941e4c9bebe4c1829b733a968bf42836/_build
Scanning dependencies of target gmock
Scanning dependencies of target gmock_main
Scanning dependencies of target gtest
[  9%] Building CXX object googletest-release-1.8.0/googlemock/CMakeFiles/gmock.dir/src/gmock-all.cc.o
[ 18%] Building CXX object googletest-release-1.8.0/googlemock/CMakeFiles/gmock.dir/__/googletest/src/gtest-all.cc.o
[ 36%] Building CXX object googletest-release-1.8.0/googlemock/CMakeFiles/gmock_main.dir/__/googletest/src/gtest-all.cc.o
[ 36%] Building CXX object googletest-release-1.8.0/googlemock/gtest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
[ 45%] Building CXX object googletest-release-1.8.0/googlemock/CMakeFiles/gmock_main.dir/src/gmock-all.cc.o
[ 54%] Linking CXX shared library ../../../lib/libgtest.so
[ 63%] Linking CXX shared library ../../lib/libgmock.so
[ 72%] Building CXX object googletest-release-1.8.0/googlemock/CMakeFiles/gmock_main.dir/src/gmock_main.cc.o
[ 72%] Built target gmock
[ 72%] Built target gtest
Scanning dependencies of target gtest_main
[ 81%] Building CXX object googletest-release-1.8.0/googlemock/gtest/CMakeFiles/gtest_main.dir/src/gtest_main.cc.o
[ 90%] Linking CXX shared library ../../../lib/libgtest_main.so
[ 90%] Built target gtest_main
[100%] Linking CXX shared library ../../lib/libgmock_main.so
[100%] Built target gmock_main
gtest/1.8.0@lasote/testing: Package '0f715b9d941e4c9bebe4c1829b733a968bf42836' built
gtest/1.8.0@lasote/testing: Build folder /home/matt/.conan/data/gtest/1.8.0/lasote/testing/build/0f715b9d941e4c9bebe4c1829b733a968bf42836
gtest/1.8.0@lasote/testing: Generated conaninfo.txt
gtest/1.8.0@lasote/testing: Generated conanbuildinfo.txt
gtest/1.8.0@lasote/testing: Generating the package
gtest/1.8.0@lasote/testing: Package folder /home/matt/.conan/data/gtest/1.8.0/lasote/testing/package/0f715b9d941e4c9bebe4c1829b733a968bf42836
gtest/1.8.0@lasote/testing: Calling package()
gtest/1.8.0@lasote/testing package(): Copied 1 '' files: LICENSE
gtest/1.8.0@lasote/testing package(): Copied 4 '.so' files: libgtest.so, libgmock_main.so, libgmock.so, libgtest_main.so
gtest/1.8.0@lasote/testing package(): Copied 41 '.h' files
gtest/1.8.0@lasote/testing: Package '0f715b9d941e4c9bebe4c1829b733a968bf42836' created
gtest/1.8.0@lasote/testing (test package): Generator cmake created conanbuildinfo.cmake
gtest/1.8.0@lasote/testing (test package): Generator txt created conanbuildinfo.txt
gtest/1.8.0@lasote/testing (test package): Generated conaninfo.txt
gtest/1.8.0@lasote/testing (test package): Running build()
ERROR: DefaultName/0.1@PROJECT: Error in build() method, line 14
	build_type = self.settings.build_type
	ConanException: 'settings.build_type' doesn't exist
'settings' possible configurations are ['arch', 'compiler', 'os']
```
Adding `-s build_type=Release` makes the build work.

**Comment by matlo607** [comment]
An interesting information. The field `build_type` was missing in my profile.
```
Configuration for profile gcc-49:

[settings]
    os=Linux
    arch=x86_64
    compiler=gcc
    compiler.version=4.9
    compiler.libcxx=libstdc++11
[options]
[build_requires]
[env]
    CFLAGS=-fpic
    CXXFLAGS=-fpic -std=c++11
```
Adding it solved the issue.
Why didn't it fail during the build of the package itself but failed during the build of test_package ?

**Comment by memsharded** [comment]
Oh, yes, that is the issue. Glad that you solved it.

> Why didn't it fail during the build of the package itself but failed during the build of test_package ?

This is indeed a very good question. I am investigating, but seems a bug.

**Comment by memsharded** [comment]
I have checked it, and it is not a bug, it is a feature! :) :)

The rationale is:
- ``build_type`` has possible values ``[None, Release, Debug]`` in *settings.yml*
- ``None`` value is special, it allows a setting not to be defined. It is used for backward compatibility when introducing new settings in recipes. These settings and recipes can be used by others without breaking and without giving values to those settings in their profiles or command line. (a different discussion is that None shouldn't be a possible value for ``build_type``, but it is there and now not easy to remove without breaking)
- *test_package/conanfile.py* is also special. It does an ``install + build + test``, so the settings state is serialized at install, and deserialized at build. So it is more restrictive, and it detects that ``build_type`` was not defined even if the *test_package* recipe is requesting to use it. So package creator is able to detect there is something wrong in their setup, while the recipe could still work for other users that do not define the setting.

I hope this makes it a bit clearer. I have added a test in https://github.com/conan-io/conan/pull/2506 that makes this behavior explicitly tested. Thanks for your feedback!
